File: webflux-api-gateway/src/main/java/tally/webflux/gateway/ApiGatewayController.java
package tally.webflux.gateway;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;
import reactor.core.publisher.Mono;

import java.time.Duration;
import java.util.concurrent.ThreadLocalRandom;

@RestController
public class ApiGatewayController {
    @GetMapping("/api/hello")
    public Mono<String> hello() {
        int delayMs = ThreadLocalRandom.current().nextInt(10, 100);
        return Mono.delay(Duration.ofMillis(delayMs))
                .thenReturn("OK - " + delayMs + "ms delay");
    }
}

File: coroutine-api-gateway/src/main/kotlin/tally/coroutine/gateway/ApiGatewayController.kt
package tally.coroutine.gateway

import kotlinx.coroutines.delay
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RestController
import kotlin.random.Random

@RestController
class ApiGatewayController {

    @GetMapping("/api/hello")
    suspend fun hello(): String {
        val delayMs = Random.nextInt(10, 100)
        delay(delayMs.toLong())
        return "OK - ${delayMs}ms delay"
    }
}